src/main/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiff.scala

```diff
@@ -70,34 +72,49 @@ object AssertOpcDiff {
-    val LteqAssert = Value(2)
-    val LtInt64Assert = Value(3)
-    val GtZeroAssert = Value(4)
-    val EqAssert = Value(5)
-    val IsCallerOriginAssert = Value(6)
-    val IsSignerOriginAssert = Value(7)
+  def checkHash(hashValue: DataEntry, hashKey: DataEntry): Either[ValidationError, OpcDiff] = {
+    if (hashValue.dataType != DataType.ShortText || hashKey.dataType != DataType.ShortText)
+      Left(ContractDataTypeMismatch)
+    else {
+      val hashResult = ByteStr(FastCryptographicHash(hashKey.data))
+      if (hashResult.equals(ByteStr(hashValue.data)))
+        Right(OpcDiff.empty)
+      else
+        Left(ContractInvalidHash)
+    }
+  }
+
+  object AssertType extends Enumeration(1) {
+    val GteqZeroAssert, LteqAssert, LtInt64Assert, GtZeroAssert, EqAssert, IsCallerOriginAssert, IsSignerOriginAssert = Value
   }
 
   def parseBytes(context: ExecutionContext)
-                (bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, OpcDiff] = bytes.head match {
-    case opcType: Byte if opcType == AssertType.GteqZeroAssert.id && bytes.length == 2
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => gtEq0(data(bytes(1)))
-    case opcType: Byte if opcType == AssertType.LteqAssert.id && bytes.length == 3
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => ltEq(data(bytes(1)), data(bytes(2)))
-    case opcType: Byte if opcType == AssertType.LtInt64Assert.id && bytes.length == 2
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => ltInt64(data(bytes(1)))
-    case opcType: Byte if opcType == AssertType.GtZeroAssert.id && bytes.length == 2
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => gt0(data(bytes(1)))
-    case opcType: Byte if opcType == AssertType.EqAssert.id && bytes.length == 3
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => eq(data(bytes(1)), data(bytes(2)))
-    case opcType: Byte if opcType == AssertType.IsCallerOriginAssert.id && bytes.length == 2
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => isCallerOrigin(context)(data(bytes(1)))
-    case opcType: Byte if opcType == AssertType.IsSignerOriginAssert.id && bytes.length == 2
-      && bytes.tail.max < data.length && bytes.tail.min >= 0 => isSignerOrigin(context)(data(bytes(1)))
-    case _ => Left(ContractInvalidOPCData)
+                (bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, OpcDiff] = {
+    if (checkAssertDataIndex(bytes, data.length)) {
+      val gteqZeroAssertId = AssertType.GteqZeroAssert.id.toByte
+      val lteqAssertId = AssertType.LteqAssert.id.toByte
+      val ltInt64AssertId = AssertType.LtInt64Assert.id.toByte
+      val gtZeroAssertId = AssertType.GtZeroAssert.id.toByte
+      val eqAssertId = AssertType.EqAssert.id.toByte
+      val isCallerOriginAssertId = AssertType.IsCallerOriginAssert.id.toByte
+      val isSignerOriginAssertId = AssertType.IsSignerOriginAssert.id.toByte
+      (bytes.headOption, bytes.length) match {
```

Comment:
How about using `AssertType(bytes.headOption)` to avoid `.id.toByte`
```scala
object Num extends Enumeration(1) { val One, Two = Value }
Some(Num(1:Byte)) match {
  case Some(Num.One) => "a"
  case Some(Num.Two) => "b"
  case _ => "e"
}
```
> res: String = a

Reply:
```scala
bytes.headOption.flatMap(f => Option(AssertType(f)))
```
is the correct one

Reply:
Oh, sorry. I mean `map` not `flatMap`, `bytes.headOption.map(AssertType(_))`

Reply:
will use
```scala
bytes.headOption.flatMap(f => Try((AssertType(f))).toOption)
```
to avoid the invalid `bytes.head` case.